### PR TITLE
Dockerize and Add Release-Push Workflow

### PR DESCRIPTION
**Pull request by lapp0**
Fixes https://github.com/outlines-dev/outlines/issues/555

- Introduces dockerized `outlines.serve.serve`.
- Introduces automatic docker image push action to `release.yml`
  - Example: [lapp0/outlines build](https://github.com/lapp0/outlines/actions/runs/7972431838/job/21764184361), [resulting image on Docker Hub](https://hub.docker.com/r/lapp0/outlines-serve)

Need some setup on your end please @rlouf:
- Let me know the outlines official docker repo URL so I can update `release.yml` and docs
- Configure the `DOCKERHUB_USERNAME` and `DOCKERHUB_TOKEN` secrets on GitHub in an environment named "release".

TODO: 
- [x] I run out of space later in the build process if I build the image. Fix it.
  - Fixed, all steps (except publish PyPi) pass https://github.com/lapp0/outlines/actions/runs/7973090033/job/21766228377
- [x] substitute official repo link in

### CODE REVIEW

**Comment by rlouf** [comment]
Thanks! Let's add a section in the docs' serve page, and maybe on the README as well?

**Comment by lapp0** [comment]
> Thanks! Let's add a section in the docs' serve page, and maybe on the README as well?

Great idea. Added to the docs with `lapp0/outlines-serve` as a placeholder for the official repo.

File: docs/quickstart.md [hunk 1]
```diff
@@ -239,15 +239,18 @@ Outlines can be deployed as a LLM service using [vLLM][vllm]{:target="_blank"} a
...
-This will by default start a server at `http://127.0.0.1:8000` (check what the console says, though)  with the OPT-125M model. If you want to specify another model:
+Or you can start the server with Outlines' official Docker image:
 
-```python
-python -m outlines.serve.serve --model="mistralai/Mistral-7B-Instruct-v0.2"
+```bash
+docker run -p 8000:8000 outlinesdev/outlines --model="mistralai/Mistral-7B-Instruct-v0.2"
 ```
 
+This will by default start a server at `http://127.0.0.1:8000` (check what the console says, though)  with the OPT-125M model. If you want to specify another model:
```
Comment (rlouf) [inline docs/quickstart.md:251]: I think this got mixed up?

File: docs/community/contribute.md [hunk 2]
```diff
@@ -42,6 +42,15 @@ pip install -e .[test]
 pre-commit install
 ```
 
+#### Developing Serve Endpoint Via Docker
```
Comment (rlouf) [inline docs/community/contribute.md:45]:
```suggestion
#### Deploying Serve Endpoint Via Docker
```

**Comment by rlouf** [comment]
This looks good to me. Is it ready on your end?

**Comment by lapp0** [comment]
@rlouf Build was successful on https://github.com/lapp0/outlines and the resulting docker container worked for me. Docs seem fine. Only question is whether it'll work with this repos configuration

We either can either release a new version to check if it works with your GitHub configuration, or I can add a `workflow_dispatch` to `release.yml` so you can run the workflow first before releasing.

**Comment by rlouf** [comment]
Good idea for the dispath! It can be a real pain to debug workflows, even more if we need to cut a new release every time there's a failure.

**Comment by rlouf** [comment]
Thanks! It would be less error-prone to have two separate `release-pypi.yml` and `release-dockerhub.yml` files instead of branching logic in the yaml.

**Comment by lapp0** [comment]
https://github.com/outlines-dev/outlines/actions/workflows/release.yml will have a "Run Workflow" button once this is in `main`. It will deploy to Docker Hub, but not perform any other actions (no PyPi).

Example run via `workflow_dispatch` https://github.com/lapp0/outlines/actions/runs/7992044914/job/21824662684

**Comment by lapp0** [comment]
Oops, looks like you already implemented. Please push yours once more.

**Comment by rlouf** [comment]
https://github.com/outlines-dev/outlines/actions/runs/7994133527/job/21831500043 ah

**Comment by lapp0** [comment]
@rlouf My push was missing a checkout step, I don't expect it to pass.

**Pull request merged**